### PR TITLE
CODAP-1446: Add Residual Plot to scatter plot Measure menu

### PR DESCRIPTION
**Pull request by bfinzer**
## Summary
- Adds a "Residual Plot" toggle to the scatter plot's Measure menu that splits the graph into a 2/3 scatter plot (upper) and a 1/3 residual plot (lower) driven by the active Movable Line, Least Squares Line, or Plotted Function.
- Consumes the split-plot infrastructure from #2659 (CODAP-1445). Depends on #2659 merging first.
- Mirrors the Squares of Residuals interaction pattern: menu item persists as user intent; when constraints break (line removed, legend added, etc.) the split tears down but the checkbox stays checked-and-grayed until reactivation.

## Feature behavior
- Split active only when: numeric x, numeric y, exactly one y attribute, no rightNumeric / topSplit / rightSplit / legend, and exactly one of {Movable Line, LSRL, Plotted Function} is visible.
- Lower plot renders one SVG circle per case at (x_screen, residual_y_screen), styled to match the upper Pixi points (legend color, blue selection fill without legend, halo when selected).
- Lower y-axis labeled "Residuals", auto-scaled to the residual range (always includes 0), non-interactive (no drag rects).
- Zero reference line at residual=0 drawn in the same style as the main plot's y=0 line.
- Hover tooltip: `<xAttrName>: <xValue>` / `Residual: <value>` — matches main-plot data-tip styling and formatting.
- Single-click on a residual point selects the case (shift/meta/ctrl extend); background click deselects.
- Selection halo syncs across both regions via the dataset.
- Upper-plot point drag: residual points reposition live (LSRL recomputes from cached values each frame).
- Marquee started in the upper plot is visually clamped to the upper region — no false promises of selecting residual points.

## Implementation notes
- Rendered as SVG rather than a second `PixiPointRenderer` (deviation from the original plan). A second Pixi renderer would have required plumbing a second WebGL context through `use-point-renderer.ts`; SVG gives correct behavior with much less surface area and adequate perf for typical scatter plots. Confirmed with author before implementation.
- New pure-function module `residual-plot-utils.ts` (23 unit tests): `getActiveLineKind`, `residualPlotIsApplicable`, `getPredictor`, `computeResiduals`, `residualDomain`.
- The `"leftLower"` `AxisPlace` from CODAP-1445 needed targeted plumbing to become first-class in the axis rendering pipeline — see `numeric-axis-helper.ts`, `use-axis.ts`, `use-sub-axis.ts`, `axis-helper.ts`, `graph-attribute-label.tsx`, `graph-axis.tsx`, `graph-content-model.ts`, and a new `static-axis-label.tsx`. Each site has a comment explaining why.
- One shared-code change: `background.tsx`'s marquee `onDrag` clamps to the upper plot region when `showLowerPlot` is true (no-op otherwise, and a no-op on map layouts).
- No V2 import/export — the feature doesn't round-trip through V2.

## Test plan
- [ ] CI green (Build, Lint, Jest, Cypress smoke, cypress/flake, S3 Deploy)
- [ ] `residual-plot-utils.test.ts` — 23 unit tests pass
- [ ] `bivariate-adornments.spec.ts` new Cypress test — full toggle lifecycle + checked-but-grayed persistence
- [ ] Manual: open a scatter plot with numeric x/y; add Movable Line, LSRL, and Plotted Function in turn; confirm Residual Plot toggles cleanly for each and that residuals track drag / line changes live
- [ ] Manual: add a legend attribute while Residual Plot is on — split tears down; remove legend — split returns
- [ ] Manual: drag a case in the upper plot with LSRL active — all residual points reposition (not just the dragged one)
- [ ] Manual: hover a residual point — tooltip shows x attribute + Residual value
- [ ] Manual: click residual point (selects), shift-click (extends), background click (deselects)
- [ ] Manual: try to marquee-drag from upper into lower — marquee visually stops at the region boundary

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

## Changes added during code review

Developed during review and pushed to this branch:

- **Decoupled residual-point selection restyle from recompute** (`1bdb726fb`) — selection changes were re-firing the full residual pipeline (incl. a fresh LSRL regression) just to update halos. Split rendering into geometry (painted under `mobx.untracked`) and an attrs-only `restyleResidualSelection`; extracted a unit-tested `residualPointStyle`.
- **Consolidated the scattered `leftLower` role checks** behind a declarative `Record<AxisPlace, IAxisPlaceTraits>` table + `getAxisPlaceTraits` (`e7bce5be1`), mirroring `PlotModel`'s property-getter pattern — a stepping stone toward axes owning their own properties. Behavior unchanged.
- **Tear down the split when the line is uncomputable** (`84dc71b31`) — e.g. an LSRL that drops below 2 finite points — instead of leaving stale residual UI on screen.
- **Copilot review fixes** (4 rounds): style-once-per-circle via D3 `.each()`; guarded the `leftLower` MultiScale writes (no per-frame domain churn during line drags); `Math.max(0, ...)` marquee-height clamp; tests keyed on the real `"__EMPTY__"` default cell key.
- **Added unit tests** for `computeResiduals`, `getPredictor` (all three line kinds + edge cases), and `residualPointStyle`; added a Cypress assertion for residual-plot selection-halo sync.

### Feature flag

The Residual Plot is gated behind the **`residualPlot`** feature flag (from the feature-flag system in #2666 / CODAP-1448, merged into this branch). The measure-menu item appears only when the flag is enabled — via `?features=residualPlot`, server config, or a document grant. Following the *gate affordances, not registrations* rule, the ScatterPlot render path keys off the persisted `showResidualPlot` state rather than the flag, so a pilot document that already has the residual plot on still opens and renders when the flag is off. (`bdea755b0`; the item's presence/absence with the flag on/off is covered by unit tests in `adornments-store-utils.test.ts`.)

### Legend support (intentionally deferred)

v1 excludes legends via `residualPlotIsApplicable`, and the residual computation is single-line and legend-blind (`Predictor = (x) => number`, using the first line / default cell). This is a **deliberate v1 simplification, not a structural limit**: per-category residuals — each point against its own category's line — are mathematically well-defined, and the adornments already store a line per cell (`lineDescriptions` / `plottedFunctions` maps, per-cell LSRL) with a case→cell mapping available. Accordingly, the residual point **styling is intentionally legend-ready** (handles category color / selection stroke) even though it's never exercised today; this is documented in `styleFor` so it isn't mistaken for dead code. Enabling legends later means dropping the exclusion and making the predictor cell-aware.

### Deferred follow-up

- **CODAP-1462** — escape user-controlled strings in graph data-tip tooltips (a pre-existing graph-wide pattern the residual tooltip shares).

### CODE REVIEW

File: v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx [hunk 1]
```diff
@@ -166,6 +174,115 @@ export const ScatterPlot = observer(function ScatterPlot({ renderer }: IPlotProp
...
+  // join, no residual/predictor recompute. This is what a selection change triggers (via
+  // refreshPointSelection) so selecting cases no longer re-runs the whole residual pipeline.
+  const applyResidualStyles = useCallback((g: SVGGElement) => {
+    select(g).selectAll<SVGCircleElement, IResidualPoint>("circle")
+      .attr("r", d => styleFor(d.caseID).radius)
+      .attr("fill", d => styleFor(d.caseID).fill)
+      .attr("stroke", d => styleFor(d.caseID).stroke)
+      .attr("stroke-width", d => styleFor(d.caseID).strokeWidth)
+      .attr("stroke-opacity", d => styleFor(d.caseID).strokeOpacity)
+  }, [styleFor])
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx:226]: Good catch — fixed by computing the style once per circle with D3 `.each()` instead of recomputing `styleFor` for each of the 5 attributes.

File: v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx [hunk 2]
```diff
@@ -166,6 +174,115 @@ export const ScatterPlot = observer(function ScatterPlot({ renderer }: IPlotProp
...
+    const xAttr = dataset?.getAttribute(xAttrID)
+    const xAttrName = xAttr?.name ?? ""
+    const float = format(".3~f")
+    const tipTextFor = (d: IResidualPoint) => {
+      const xValue = dataset?.getValue(d.caseID, xAttrID)
+      const xValueStr = typeof xValue === "number" && isFinite(xValue)
+        ? float(xValue) : (xValue != null ? String(xValue) : "")
+      const residualStr = isFinite(d.residual) ? float(d.residual) : String(d.residual)
+      return `${xAttrName}: ${xValueStr}<br/>${t("V3.ResidualPlot.dataTip.residual")}: ${residualStr}`
+    }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx:261]: You're right that the residual tooltip interpolates user-controlled attribute name/value into HTML without escaping. This mirrors the main graph point tooltip (`caseTipText` in `data-display-content-model.ts`), which builds the same unescaped `${name}: ${value}` string and renders it via d3-tip `.html()` — so it's an existing graph-wide pattern, not something new here. We'd rather fix it consistently across all graph tooltips than spot-fix just this one (which would give false assurance while the main tooltip stays unescaped), so we're treating it as a separate follow-up.

File: v3/src/components/graph/plots/scatter-plot/residual-plot-utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+import { kMovableLineType } from "../../adornments/movable-line/movable-line-adornment-types"
+import { kPlottedFunctionType } from "../../adornments/plotted-function/plotted-function-adornment-types"
+import { IGraphDataConfigurationModel } from "../../models/graph-data-configuration-model"
+import {
+  defaultSelectedColor, defaultSelectedStroke, defaultSelectedStrokeOpacity, defaultSelectedStrokeWidth,
+  defaultStrokeOpacity, defaultStrokeWidth
+} from "../../../../utilities/color-utils"
+import {
+  getActiveLineKind, IResidualPoint, residualDomain, residualPlotIsApplicable, residualPointStyle
+} from "./residual-plot-utils"
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/residual-plot-utils.test.ts:13]: Good call. Added focused unit tests for `computeResiduals` (residual math + filtering of non-finite x/y/predicted + the no-dataset case) and `getPredictor` (movable line, fresh LSRL recompute, plotted function including the throw→NaN path, and the null / insufficient-data cases).

File: v3/src/components/data-display/components/background.tsx [hunk 4]
```diff
@@ -131,11 +131,25 @@ export const Background = forwardRef<SVGGElement | HTMLDivElement, IProps>((prop
...
+    // Coord system: marqueeState.y is plot-area-relative (0 = top of upper region), so the upper
+    // region's bottom edge is at y = graphLayout.plotHeight.
+    let clampedHeight = height.current
+    if (graphLayout?.showLowerPlot) {
+      const plotBottom = graphLayout.plotHeight
+      const rectBottom = startY.current + height.current
+      if (height.current > 0 && rectBottom > plotBottom) {
+        clampedHeight = plotBottom - startY.current
+      }
+    }
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/data-display/components/background.tsx:151]: Good catch — clamped with `Math.max(0, plotBottom - startY.current)` so the marquee rect height can't go negative when a drag starts at or below the upper region's bottom edge.

File: v3/src/components/graph/plots/scatter-plot/residual-plot-utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+      pf: true,
+      pfModel: {
+        instanceKey: () => "{}",
+        plottedFunctions: new Map([["{}", { formulaFunction: () => { throw new Error("boom") } }]])
+      }
+    })
+    const predictor = getPredictor(store, fakeConfigWithData([]))
+    expect(predictor).not.toBeNull()
+    expect(Number.isNaN(predictor?.(1) as number)).toBe(true)
+  })
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/residual-plot-utils.test.ts:311]: Good call — updated the plotted-function tests to stub `instanceKey` → `"__EMPTY__"` and key the map with `"__EMPTY__"` (the real default cell key), so they now fail if production ever hard-codes `"{}"` instead of calling `instanceKey({})`.

File: v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx [hunk 6]
```diff
@@ -332,6 +465,61 @@ export const ScatterPlot = observer(function ScatterPlot({ renderer }: IPlotProp
...
+          axis.setAllowRangeToShrink(true)
+          axis.setDomain(minY, maxY)
+        }
+      }
+      // The graph-controller's installAxisReaction only watches "left" and "bottom"; changes to the
+      // "leftLower" axis model do not automatically push through to the MultiScale that owns the D3
+      // scale used for pixel mapping. Sync explicitly.
+      layout.setAxisScaleType("leftLower", "linear")
+      layout.getAxisMultiScale("leftLower").setNumericDomain([minY, maxY])
+      renderResidualPoints(residuals)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx:538]: Good catch — this was the one unguarded write in the reaction. Guarded both: `setAxisScaleType` now runs only when the scale type isn't already `linear`, and `setNumericDomain` only when the MultiScale's current domain differs from `[minY, maxY]`. Guarding on the MultiScale's own domain (not the axis model's) so the initial sync still happens on first activation, when the model already carries the domain. This matches the "guarded mutations" invariant and removes the per-frame domain churn during line drags.

File: v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx [hunk 7]
```diff
@@ -332,6 +465,69 @@ export const ScatterPlot = observer(function ScatterPlot({ renderer }: IPlotProp
...
+    return mstAutorun(() => {
+      const isActive = adornmentsStore.showResidualPlot && residualPlotIsApplicable(adornmentsStore, dataConfiguration)
+      if (!isActive) {
+        if (layout.showLowerPlot) layout.setShowLowerPlot(false)
+        if (graphModel.getAxis("leftLower")) graphModel.removeAxis("leftLower")
+        if (residualPointsRef.current) select(residualPointsRef.current).selectAll("circle").remove()
+        return
+      }
+      const predictor = getPredictor(adornmentsStore, dataConfiguration)
+      if (!predictor || !dataConfiguration) return
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/plots/scatter-plot/scatter-plot.tsx]: Good catch — real stale-UI edge. Extracted the teardown (collapse lower region, drop the `leftLower` axis, clear the points) into an idempotent `teardown()` and now call it from both early returns, so an uncomputable line (e.g. an LSRL that drops below 2 finite points, or a not-yet-populated plotted function) tears the split down instead of leaving stale residual UI on screen.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Looks good -- approving in principle, but requesting changes since this PR should not be merged until after the feature flag system (PR #2666) is merged and the residual plot can be gated behind a feature flag. The following review, developed in conjunction with Claude Code 🤖, describes some changes that Claude and I made on review. Look them over before merging (once the feature flag is implemented).

## PR Review Summary — #2663 (CODAP-1446)

**Changes:** 18 files (original) + review fixes. Base: `CODAP-1445-split-plot-infrastructure` (#2659).

### What it does

Adds a "Residual Plot" toggle to the scatter plot's Measure menu. When on (and applicable), the
graph splits into a 2/3 upper scatter plot and a 1/3 lower residual plot driven by the active
Movable Line, LSRL, or Plotted Function. Consumes the split-plot infrastructure from #2659. Follows
the Squares-of-Residuals pattern: the checkbox persists as user intent and stays checked-but-grayed
when constraints break.

### The approach

- **`residual-plot-utils.ts`** — pure functions (`getActiveLineKind`, `residualPlotIsApplicable`,
  `getPredictor`, `computeResiduals`, `residualDomain`, `residualPointStyle`), one shared
  applicability predicate used by both menu gating and render gating.
- **`ScatterPlot`** renders residual points as **SVG circles** (a deliberate deviation from the
  planned second Pixi renderer — confirmed with author). A `syncResidualPlot` mstAutorun owns the
  derived state: activates the split, registers/updates the `leftLower` NumericAxisModel with the
  auto-scaled residual domain, and paints points. Live drag flows through `refreshPointPositions`;
  selection halo through `refreshPointSelection`.
- **`leftLower` axis pipeline plumbing** across the axis rendering path plus a new non-interactive
  `StaticAxisLabel`.
- **`background.tsx`** marquee visually clamps to the upper region when the split is active.

### Assessment

Strong, carefully-reasoned PR. High-risk areas were verified during review:

- **#2659 cross-PR watch item resolved** — `rescale()`, `growNumericAxesToFit()`, and
  `use-sub-axis` all skip adornment-owned axes, so the externally-set residual domain is never
  clobbered by the y-attribute's extent.
- **Marquee refactor is behavior-preserving** for normal graphs; the split clamp is purely visual
  (residual points aren't in the selection R-tree).
- **No infinite-reaction loop** in `syncResidualPlot`; MultiScale setters are guarded.
- **Serialization safe** — `leftLower` axis round-trips in V3; the V2 exporter reads axes by
  explicit key and ignores it.

**Design discussion — SVG vs a second Pixi renderer:** the choice is sound. V3's WebGL contexts are
a hard-capped pooled resource (`DEFAULT_MAX_CONTEXTS = 14` across the whole document); a second Pixi
renderer per graph would double-consume that pool and fight the very system built to manage it. The
residual plot is a low-frequency, opt-in secondary view, so SVG's per-node cost is acceptable at
classroom scale, and SVG buys free per-point hit-testing / tooltips and drops into the existing D3
adornment overlay. The genuine risk is the drag-repaint path on very large datasets, which is
bounded by the applicability constraints and "typical scatter plot" scale.

### Changes made during review (committed to the branch)

The following were developed collaboratively during this review and pushed to the PR branch:

1. **Decoupled residual-point selection restyle from recompute** (`1bdb726fb`). The
   `syncResidualPlot` autorun painted via a function that read `isCaseSelected`, so every selection
   change re-fired the whole pipeline — including a full LSRL regression — purely to update halos,
   and `refreshPointSelection` redundantly re-ran it too. Split rendering into geometry
   (`renderResidualPoints`, applying selection styling under `mobx.untracked` so the autorun no
   longer subscribes to selection) and an attrs-only `restyleResidualSelection`. Extracted the pure
   style decision into `residualPointStyle` (unit-tested).

2. **Copilot fixes, round 1** (`f6c31d96c`): `applyResidualStyles` computes the style once per
   circle via D3 `.each()`; added unit tests for `computeResiduals` and `getPredictor` (all three
   line kinds + edge cases).

3. **Copilot fixes, round 2** (`430a6dfff`): `background.tsx` clamps the marquee height with
   `Math.max(0, …)` (no negative SVG rect); the plotted-function tests key on the real default cell
   key `"__EMPTY__"`.

4. **Copilot fix, round 3** (`6e6bc4b6a`): guarded the `leftLower` MultiScale `setAxisScaleType` /
   `setNumericDomain` writes so a re-fire with an unchanged scale type / domain is a no-op (removes
   per-frame domain churn during line drags), honoring the reaction's documented guarded-mutations
   invariant.

5. **Axis-place trait table** (`e7bce5be1`): consolidated the scattered `place === 'leftLower'`
   *role* checks behind a declarative `Record<AxisPlace, IAxisPlaceTraits>` (`isAdornmentOwned`,
   `isInteractive`, `fixedLabelKey`, `alwaysShowsZeroLine`) + `getAxisPlaceTraits`, mirroring
   `PlotModel`'s property-getter pattern. 9 call sites now query intent instead of the place string;
   place-intrinsic geometry stays in its existing per-place tables. A stepping stone toward
   axis-owned properties; behavior unchanged.

6. **Copilot fix, round 4** (`84dc71b31`): `syncResidualPlot` now tears down the split when
   `getPredictor()` returns null (e.g. an LSRL that drops below 2 finite points, or a not-yet-
   populated plotted function) while applicability still holds, instead of leaving stale residual UI
   on screen. Teardown extracted to an idempotent helper called from both early returns.

Also: added a Cypress assertion for residual-plot selection-halo sync (with a `{ force: true }`
click, since residual points for near-identical residuals overlap).

### Issues

No open blocking issues. Verification: **39 residual-plot-utils unit tests + 104 graph/axis unit
tests + full `build:tsc` + lint all green locally**; a full-lifecycle Cypress test plus the new
selection assertion cover the feature end-to-end.

Deferred / tracked:

- **Tooltip HTML escaping** (Copilot) — the residual tooltip interpolates user-controlled attribute
  name/value into HTML via d3-tip `.html()`. This is a **pre-existing, graph-wide pattern** (the
  main point tooltip `caseTipText` is identical), so rather than spot-fix one tooltip it's tracked
  for a consistent graph-wide fix in **CODAP-1462**.
- **`caseIds` delta on `restyleResidualSelection`** — a Pixi-style delta-restyle could make marquee
  selection O(delta) instead of O(N); deferred as a future optimization (residual points restyle all
  circles today, same as before this change).
- **Legend support (per-category residuals)** — v1 excludes legends via `residualPlotIsApplicable`,
  and the residual *computation* is single-line and legend-blind: `getPredictor` collapses to the
  first line / default cell, and the `Predictor` type is `(x) => number` (no case/category context).
  But this is a **deliberate v1 simplification, not a structural limit** — per-category residuals
  (each point against its own category's line) are mathematically well-defined, and the adornments
  already store a line per cell (`lineDescriptions` map, `plottedFunctions` map, per-cell LSRL) with
  a case→cell mapping available. Consequently the residual point **styling is intentionally
  legend-ready** (handles category color / selection stroke) even though it's never exercised today;
  this is now documented in `styleFor` (commit `b7864869c`) so it doesn't read as dead code. Enabling
  legends later means dropping the exclusion and making the predictor cell-aware — no ticket filed
  (no current plan). Flagged so the styling isn't mistaken for an oversight.

### Note for the author

Substantial changes were made to this PR during review (the selection-restyle decouple, the
axis-place trait table, four rounds of Copilot fixes, and the uncomputable-line teardown) — Bill
should review the scope of what changed before it's marked ready. One design point worth his eye:
the **styling anticipates legends while the computation precludes them** (see the legend note above).
It's now documented as an intentional v1 staging, but he should confirm that matches his intent.

**Review by kswenson — APPROVED** [review]
Approving now that the residual plot UI is feature-flag-gated.
